File: Kadane Algorithm/kadaneAlgo.py
def kadane_algorithm(arr):
    max_sum = arr[0] # Initialize a variable to store the maximum subarray sum
    current_sum = 0
    
    for num in arr:
        current_sum += num # Add the current element to the current subarray sum
        max_sum = max(max_sum, current_sum) # Update the maximum subarray sum if the current sum is greater
        
        if current_sum < 0:
            current_sum = 0 # If the current subarray sum becomes negative, reset it to 0
    
    return max_sum

if __name__ == "__main__":
    arr = [-2, 1, -3, 4, -1, 2, 1, 5, 4]
    max_sum = kadane_algorithm(arr) # Call the Kadane's Algorithm function
    print("Maximum subarray sum:", max_sum)

File: Kadane Algorithm/kadaneAlgo.java
public class kadaneAlgo {

    // Function to find the maximum subarray sum using Kadane's Algorithm
    public static int kadaneAlgorithm(int[] arr) {
        int maxSum = arr[0]; // Initialize a variable to store the maximum subarray sum
        int currentSum = 0;

        for (int i = 0; i < arr.length; i++) {
            currentSum += arr[i]; // Add the current element to the current subarray sum
            maxSum = Math.max(maxSum, currentSum); // Update the maximum subarray sum if the current sum is greater

            if (currentSum < 0) {
                currentSum = 0; // If the current subarray sum becomes negative, reset it to 0
            }
        }

        return maxSum;
    }

    public static void main(String[] args) {
        int[] arr = {-2, 1, -3, 4, -1, 2, 1, 5, 4};
        int maxSum = kadaneAlgorithm(arr); // Call the Kadane's Algorithm function
        System.out.println("Maximum subarray sum: " + maxSum);
    }
}
